### PR TITLE
80x new discriminators

### PR DESCRIPTION
**Pull request by andrewj314**
Updated tau ID discriminators for 13 TeV & MiniAODv2: updated anti-electron ID from MVA5 to MVA6, added new ID discriminators for 13TeV, removed obsolete discriminators, and updated the db used from CMS_COND_PAT_000 to CMS_CONDITIONS - redo of PR #12298 for 80x

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@andrewj314: 
There are static analyzer warnings to fix on PFRecoTauClusterVariables.h and AntiElectronIDMVA6.cc.
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12468/9799/llvm-analysis/

The unit test failure seems to have something to do with taus:

   [1] Running path 'outpath'
   [2] Calling event method for module PoolOutputModule/'out'
   [3] Calling produce method for unscheduled module PATTauSelector/'selectedPatTaus'
   [4] Calling produce method for unscheduled module PATTauProducer/'patTaus'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: reco::PFTauDiscriminator
Looking for module label: hpsPFTauDiscriminationByMVA6LooseElectronRejection
Looking for productInstanceName: 

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "SkipEvent = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------

**Comment by cvuosalo** [comment]
@andrewj314: Why do the Python files have so much commented-out code? Old versions are preserved in Git, so why are you keeping the old lines in the new versions of the files?

File: RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    return tau.isolationPFGammaCands();
+  }
+
+  bool isInside(float photon_pt, float deta, float dphi){
+
+    if(photon_pt==0){return false;}
+
+    if(
+       (dphi < TMath::Min(0.3, TMath::Max(0.05, 0.352476*TMath::Power(photon_pt, -0.707716)))) && \
+       (deta < TMath::Min(0.15, TMath::Max(0.05, 0.197077*TMath::Power(photon_pt, -0.658701))))
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h]: Please no magic numbers. Set well-named constants to these values and then use them in place of the magic numbers.

Comment (slava77) [inline RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h]: please remove use cases of TMath where std:: equivalents are available
std::min, std::max, std::pow are available in all cases for these lines.
The are more cases like this in other files of this PR.

File: RecoTauTag/RecoTau/interface/AntiElectronIDMVA6.h [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+   Float_t* Var_wGwGSF_Endcap_;
+   
+   const GBRForest* mva_NoEleMatch_woGwoGSF_BL_;
+   const GBRForest* mva_NoEleMatch_wGwoGSF_BL_;
+   const GBRForest* mva_woGwGSF_BL_;
+   const GBRForest* mva_wGwGSF_BL_;
+   const GBRForest* mva_NoEleMatch_woGwoGSF_EC_;
+   const GBRForest* mva_NoEleMatch_wGwoGSF_EC_;
+   const GBRForest* mva_woGwGSF_EC_;
+   const GBRForest* mva_wGwGSF_EC_;
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/interface/AntiElectronIDMVA6.h]: Please don't use bare pointers like Float_t* and GBRForest*. I think the Var_ members are intended to be arrays, so you can use std::vector, while the mva_ members can be unique_ptr.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 3]
```diff
@@ -0,0 +1,965 @@
...
+  mvaName_wGwGSF_EC_ = cfg.getParameter<std::string>("mvaName_wGwGSF_EC");
+
+  Var_NoEleMatch_woGwoGSF_Barrel_ = new Float_t[10];
+  Var_NoEleMatch_wGwoGSF_Barrel_ = new Float_t[18];
+  Var_woGwGSF_Barrel_ = new Float_t[24];
+  Var_wGwGSF_Barrel_ = new Float_t[32];
+  Var_NoEleMatch_woGwoGSF_Endcap_ = new Float_t[9];
+  Var_NoEleMatch_wGwoGSF_Endcap_ = new Float_t[17];
+  Var_woGwGSF_Endcap_ = new Float_t[23];
+  Var_wGwGSF_Endcap_ = new Float_t[31];
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: More magic numbers here. Please replace them with constants. You don't even need them here if you use std::vector for Var_.

File: RecoTauTag/RecoTau/interface/AntiElectronIDMVA6.h [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+   std::vector<TFile*> inputFilesToDelete_;
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/interface/AntiElectronIDMVA6.h]: inputFilesToDelete_ should contain unique_ptr, not bare TFile pointers.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 5]
```diff
@@ -0,0 +1,965 @@
...
+    double eta_i = GammasdEtaInSigCone[i];
+    sumPt  +=  pt_i;
+    sumPt2 += (pt_i*pt_i);
+    dEta   += (pt_i*eta_i);
+    dEta2  += (pt_i*eta_i*eta_i);
+    dPhi   += (pt_i*phi_i);
+    dPhi2  += (pt_i*phi_i*phi_i);
+  }
+
+  Float_t TauGammaEnFracIn = sumPt/TauPt;
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: Please check TauPt > 0 before dividing, just like you do for sumPt in the next line.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 6]
```diff
@@ -0,0 +1,965 @@
...
+  else if ( TauSignalPFGammaCandsIn == 0 && TauHasGsf > 0.5 ) {
+    if ( std::abs(TauEtaAtEcalEntrance) < 1.479 ) {
+      Var_woGwGSF_Barrel_[0]  = std::max(float(-0.1), ElecEtotOverPin);
+      Var_woGwGSF_Barrel_[1]  = std::log(ElecChi2NormGSF);
+      Var_woGwGSF_Barrel_[2]  = ElecGSFNumHits;
+      Var_woGwGSF_Barrel_[3]  = std::log(ElecGSFTrackResol);
+      Var_woGwGSF_Barrel_[4]  = ElecGSFTracklnPt;
+      Var_woGwGSF_Barrel_[5]  = ((ElecGSFNumHits - ElecKFNumHits)/(ElecGSFNumHits + ElecKFNumHits));
+      Var_woGwGSF_Barrel_[6]  = std::log(ElecChi2NormKF);
+      Var_woGwGSF_Barrel_[7]  = std::min(std::abs(ElecPin - ElecPout)/ElecPin, float(1.));
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: Check (ElecPin > 0) before dividing.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 7]
```diff
@@ -0,0 +1,965 @@
...
+      Var_wGwGSF_Barrel_[30] = TaudCrackEta;
+      Var_wGwGSF_Barrel_[31] = TaudCrackPhi;
+      mvaValue = mva_wGwGSF_BL_->GetClassifier(Var_wGwGSF_Barrel_);
+    } else {
+      Var_wGwGSF_Endcap_[0]  = std::max(float(-0.1), ElecEtotOverPin);
+      Var_wGwGSF_Endcap_[1]  = std::log(ElecChi2NormGSF);
+      Var_wGwGSF_Endcap_[2]  = ElecGSFNumHits;
+      Var_wGwGSF_Endcap_[3]  = std::log(ElecGSFTrackResol);
+      Var_wGwGSF_Endcap_[4]  = ElecGSFTracklnPt;
+      Var_wGwGSF_Endcap_[5]  = ((ElecGSFNumHits - ElecKFNumHits)/(ElecGSFNumHits + ElecKFNumHits));
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: Need more checks against division by 0: here and lines 399, 422, and 450.
Also, please replace magic number 1.479 with constant throughout.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 8]
```diff
@@ -0,0 +1,965 @@
...
+  std::vector<Float_t> GammasPtOutSigCone;
+  reco::Candidate::LorentzVector pfGammaSum(0,0,0,0);
+  reco::Candidate::LorentzVector pfChargedSum(0,0,0,0);
+  
+  for ( unsigned i = 0 ; i < thePFTau.signalPFGammaCands().size(); ++i ) {
+    reco::PFCandidatePtr gamma = thePFTau.signalPFGammaCands().at(i);
+    float dR = deltaR(gamma->p4(), thePFTau.leadPFChargedHadrCand()->p4());
+
+    // pfGammas inside the tau signal cone
+    if (dR < std::max(0.05, std::min(0.10, 3.0/thePFTau.pt()))) {
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: Please check thePFTau.pt() > 0.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 9]
```diff
@@ -0,0 +1,965 @@
...
+	  pfCluster != pfSuperCluster->clustersEnd(); ++pfCluster ) {
+      double pfClusterEn = (*pfCluster)->energy();
+      if ( pfCluster == pfSuperCluster->clustersBegin() ) ElecEe += pfClusterEn;
+      else ElecEgamma += pfClusterEn;
+    }
+  }
+  
+  Float_t ElecPin = std::sqrt(theGsfEle.trackMomentumAtVtx().Mag2());
+  Float_t ElecPout = std::sqrt(theGsfEle.trackMomentumOut().Mag2());
+  Float_t ElecEtotOverPin = (ElecEe + ElecEgamma)/ElecPin;
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: Check ElecPin > 0.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 10]
```diff
@@ -0,0 +1,965 @@
...
+  std::vector<Float_t> GammasPtOutSigCone;
+  reco::Candidate::LorentzVector pfGammaSum(0,0,0,0);
+  reco::Candidate::LorentzVector pfChargedSum(0,0,0,0);
+  
+  for ( unsigned i = 0 ; i < thePFTau.signalPFGammaCands().size(); ++i ) {
+    reco::PFCandidatePtr gamma = thePFTau.signalPFGammaCands().at(i);
+    float dR = deltaR(gamma->p4(), thePFTau.leadPFChargedHadrCand()->p4());
+
+    // pfGammas inside the tau signal cone
+    if (dR < std::max(0.05, std::min(0.10, 3.0/thePFTau.pt()))) {
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: Check thePFTau.pt() > 0.

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 11]
```diff
@@ -0,0 +1,965 @@
...
+}
+
+double AntiElectronIDMVA6::minimum(double a, double b)
+{
+  if ( std::abs(b) < std::abs(a) ) return b;
+  else return a;
+}
+
+
+#include<array>
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc]: Why is this #include here? It should be at the top of the file with the rest so readers can easily find it.

File: RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    Float_t total = EcalEnInSignalPFCands + HcalEnInSignalPFCands;
+    if(total==0) return -1;
+    else return EcalEnInSignalPFCands/total;
+  }
+
+  float pt_weighted_dx(const reco::PFTau& tau, int mode = 0, int var = 0, int decaymode = -1){
+  
+    float sum_pt = 0.;
+    float sum_dx_pt = 0.;
+    float signalrad = std::max(0.05, std::min(0.1, 3./tau.pt()));
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h]: Check tau.pt() > 0 before dividing.

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationAgainstElectronMVA6.cc [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+    int numSignalPFGammaCandsInSigCone = 0;
+    const std::vector<reco::PFCandidatePtr>& signalPFGammaCands = thePFTauRef->signalPFGammaCands();
+    
+    for ( std::vector<reco::PFCandidatePtr>::const_iterator pfGamma = signalPFGammaCands.begin();
+	  pfGamma != signalPFGammaCands.end(); ++pfGamma ) {
+            
+      double dR = deltaR((*pfGamma)->p4(), thePFTauRef->leadPFChargedHadrCand()->p4());
+            
+      // pfGammas inside the tau signal cone
+      if (dR < std::max(0.05, std::min(0.10, 3.0/thePFTauRef->pt()))) {
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationAgainstElectronMVA6.cc]: Check thePFTauRef->pt() > 0.

File: RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationAgainstElectronMVA6.cc [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+	  //// Veto taus that go to Ecal crack
+	  if ( isInEcalCrack(tauEtaAtEcalEntrance) || isInEcalCrack(leadChargedPFCandEtaAtEcalEntrance) ) {
+	    // add category index
+	    category_output_->setValue(tauIndex_, category);
+	    // return MVA output value
+	    return -99;
+	  }
+	  //// Veto taus that go to Ecal crack
+
+	  if ( TMath::Abs(tauEtaAtEcalEntrance) < 1.479 ) { // Barrel
```
Comment (cvuosalo) [inline RecoTauTag/RecoTau/plugins/PFRecoTauDiscriminationAgainstElectronMVA6.cc]: Replace TMath with std.

**Comment by veelken** [comment]
Hi Carl,

I have checked the usage of the bare pointers of type Float_t* and GBRForest* 
in RecoTauTag/RecoTau/interface/AntiElectronIDMVA6.h . I don't see an easy way to avoid using these bare pointers: Float_t* is needed because GBRForest->GetClassifier() expexts as bare pointer to an array of floats as input (see CondFormats/EgammaObjects/interface/GBRForest.h). The bare pointers to GBRForest are used because they allow us to load the GBRForest objects either from a ROOT file or from the database - and only in case the GBRForest objects are loaded from a ROOT file, delete needs to be called for the GBRForest* pointers.

We would need an additional function GBRForest->GetClassifier(const std::vector<float>&) in the GBRForest class first before we can avoid the usage of bare pointers of type Float_t* in RecoTauTag/RecoTau/interface/AntiElectronIDMVA6.h . 

In case you have a proposal how to change the code such that it support loading the GBRForest objects from a ROOT file and from the database without using bare pointers of type GBRForest*, please let me know.

Thank you very much,


Christian

**Comment by ferencek** [comment]
@veelken, try using the TMVAEvaluator (`CommonTools/Utils/interface/TMVAEvaluator.h`, `CommonTools/Utils/src/TMVAEvaluator.cc`) or borrow some of its GBRForest implementation details.

**Comment by cvuosalo** [comment]
@veelken: I think fixing the bare pointers should be pretty easy. std::shared_ptr() might be more appropriate instead of unique_ptr() if the objects are managed by other code. The code that @ferencek mentioned shows how to create the float arrays:

std::unique_ptr<float[]> vars(new float[array_size]);

If it is a big effort to use smart pointers with GBRForest calls, we can leave it to a later PR, but I think it should be fairly simple, since it has already been done in the TMVAEvaluator code.

**Pull request merged**